subprojects/core/src/integTest/groovy/org/gradle/api/HttpScriptPluginIntegrationSpec.groovy

```diff
@@ -65,6 +66,96 @@ class HttpScriptPluginIntegrationSpec extends AbstractIntegrationSpec {
         "https" | true
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/2891")
+    def "can apply script with URI containing a query string"() {
+        when:
+        def queryString = 'p=foo;a=blob_plain;f=bar;hb=foo/bar/foo'
+        def script = file('external.gradle')
+        server.expectGetWithQueryString('/external.gradle', queryString, script)
+
+        script << """
+            task doStuff
+            assert buildscript.sourceFile == null
+            assert "${server.uri}/external.gradle?$queryString" == buildscript.sourceURI as String
+"""
+
+        buildFile << """
+            apply from: '$server.uri/external.gradle?$queryString'
+            defaultTasks 'doStuff'
+"""
+
+        then:
+        succeeds()
+    }
+
+    def "scripts with same URI path but different query strings are treated as separate things"() {
+        when:
+        def first = file('first.gradle') << """
+            task first
+            assert buildscript.sourceFile == null
+            assert "${server.uri}/external.gradle?first" == buildscript.sourceURI as String
+        """
+        def second = file('second.gradle') << """
+            task second
+            assert buildscript.sourceFile == null
+            assert "${server.uri}/external.gradle?second" == buildscript.sourceURI as String
+        """
+        server.expectGetWithQueryString('/external.gradle', "first", first)
+        server.expectGetWithQueryString('/external.gradle', "second", second)
+
+        buildFile << """
+            apply from: '$server.uri/external.gradle?first'
+            apply from: '$server.uri/external.gradle?second'
+            defaultTasks 'first', 'second'
+"""
+
+        then:
+        succeeds()
+    }
+
+    def "does not cache URIs with query parts"() {
+        when:
+        def queryString = 'p=foo;a=blob_plain;f=bar;hb=foo/bar/foo'
+        def script = file('external.gradle')
+        server.expectGetWithQueryString('/external.gradle', queryString, script)
+
+        script << """
+            task doStuff
+            assert buildscript.sourceFile == null
+            assert "${server.uri}/external.gradle?$queryString" == buildscript.sourceURI as String
+"""
+
+        buildFile << """
+            apply from: '$server.uri/external.gradle?$queryString'
+            defaultTasks 'doStuff'
```

Comment:
I don't think there's a real need to create a task for the scenario. We could just call `help` or `tasks` I guess.